core/templates/pages/contributor-dashboard-page/services/contribution-and-review.service.spec.ts

```diff
@@ -311,11 +313,41 @@ describe('Contribution and review service', () => {
           '2022-01-02'
         )
         .then(response => {
-          expect(response.from_date).toEqual('1 Nov 2022');
-          expect(response.to_date).toEqual('1 Dec 2022');
-          expect(response.contribution_hours).toEqual(1.0);
-          expect(response.team_lead).toEqual('Test User');
-          expect(response.language).toEqual('Hindi');
+          const data = response.certificate_data;
+
+          if (!data) {
+            fail('expected data to not be null');
+            return;
+          }
+
+          expect(data.from_date).toEqual('1 Nov 2022');
+          expect(data.to_date).toEqual('1 Dec 2022');
+          expect(data.contribution_hours).toEqual(1.0);
+          expect(data.team_lead).toEqual('Test User');
+          expect(data.language).toEqual('Hindi');
+        });
+
+      expect(downloadContributorCertificateAsyncSpy).toHaveBeenCalled();
+    });
+
+    it('should return null when contributor certificate data is unavailable', () => {
```

Comment:
You probably need to have an "async" before the function since you have long-running operations in this test? Ditto above. See e.g. line 335 of translate-text-backend-api.service.spec.ts.

In its current state, does the test actually fail if your Promise.resolve() in line 335 doesn't resolve to null?

Reply:
That is a good observation! Currently, the test does fail when it is supposed to, i.e., when the promise is resolved with data rather than null.

Looking at the asynchronous tests in general, it seems there are a few different formats (line numbers are from contribution-and-review.service.spec.ts):

1. Some pass an async function to "it" and use "await" (line 470)
2. Some pass an async function and use "await" with "expectAsync" (lines 280 and 309 from translate-text-backend-api.service.spec.ts)
3. some pass an async function but don't use await/return values (lines 437)
4. others pass a synchronous function and put the 'expect' expressions within the callback to "then" (lines 294, 333, 358, 383, 418)
5. still others use fakeAsync so they can sue the tick()/flushMicrotasks() methods (lines 166, 241, 739, 768, 815, 844, 887, 912, 1023, 1051, most other tests in translate-text-backend-api.service.spec.ts)

The tests relevant to this PR belong to case 4 listed above. I can change just those tests, or all of the tests of the 4th case, or just keep it as it is. I'm not sure which ways are considered more standard practice (cases 1-4 seem more or less interchangeable (?), obviously case 5 is different because more fine-grained control was needed).

Reply:
@seanlip It looks like you were on to something! Although the test for expecting null works as expected, there is indeed an issue with our asynchronous tests. It seems the tests in case 4 mentioned in my previous post do not actually run the expect clauses in their <ode>then</code> callback functions.

Like you mentioned, one way to fix this is by adding <code>async</code>; however, apparently we must also add <code>await</code> to the asynchronous operation we are testing so that the test knows to wait for the promise to resolve.

Since this issue may affect multiple files outside of this PR, I opened a new issue about it (#20317).

Reply:
Ok, for now I added async to the test callback function, and await on for the asynchronous operation. This gives the desired behavior in all circumstances so far.